hasher.go

```diff
@@ -159,21 +164,42 @@ func CalculateLimit(maxCapacity, numItems, size uint64) uint64 {
 
 func (h *Hasher) FillUpTo32() {
 	// pad zero bytes to the left
-	if rest := len(h.buf) % 32; rest != 0 {
-		h.buf = append(h.buf, zeroBytes[:32-rest]...)
+	if h.pos != 0 {
+		copy(h.buf[len(h.buf)-1][h.pos:], zeroBytes[:32-h.pos])
 	}
+	h.pos = 0
 }
 
 func (h *Hasher) AppendUint8(i uint8) {
-	h.buf = MarshalUint8(h.buf, i)
+	dst := make([]byte, 0)
+	dst = MarshalUint8(dst, i)
+	if h.pos == 0 {
+		var b32 [32]byte
+		copy(b32[:], dst)
+		h.buf = append(h.buf, b32)
+	} else {
+		copy(h.buf[len(h.buf)-1][h.pos:], dst)
+	}
+	h.pos = (h.pos + 1) % 32
 }
 
 func (h *Hasher) AppendUint64(i uint64) {
-	h.buf = MarshalUint64(h.buf, i)
+	dst := make([]byte, 0)
+	dst = MarshalUint64(dst, i)
+	willFit := h.pos <= 24
+	if willFit {
+		copy(h.buf[len(h.buf)-1][h.pos:], dst)
+	} else {
+		copy(h.buf[len(h.buf)-1][h.pos:], dst[:32-h.pos])
+		var b32 [32]byte
+		copy(b32[:], dst[32-h.pos:])
+		h.buf = append(h.buf, b32)
+	}
+	h.pos = (h.pos + 8) % 32
 }
 
-func (h *Hasher) Append(i []byte) {
-	h.buf = append(h.buf, i...)
+func (h *Hasher) Append(i [32]byte) {
+	h.buf = append(h.buf, i)
```

Comment:
are you sure about this? Don't you need to deal with chunks like in all of the above?